src/auth_middleware.rs

```diff
@@ -0,0 +1,132 @@
+use hmac::Mac;
+use http_body_util::combinators::UnsyncBoxBody;
+use std::mem;
+use std::task::{self, Poll};
+
+use anyhow::Context;
+use bytes::Bytes;
+use futures_util::future::BoxFuture;
+use hmac::Hmac;
+use http_body_util::{BodyExt, Full};
+use sha2::Sha256;
+use tower::{Layer, Service};
+
+use crate::config::Config;
+
+type Hmac256 = Hmac<Sha256>;
+
+#[derive(Debug, Clone, Default)]
+pub struct AuthLayer {
+    pub shared_secret: Vec<String>,
+}
+
+// TODO do I need the layer?
```

Comment:
Do you?

Reply:
Yes, we need a Layer as the AuthService can't be connected to the server directly as it doesn't have any routes.